contrib/pinot-druid-migration/src/main/java/org/apache/pinot/druid/data/readers/DruidSegmentRecordReader.java

```diff
@@ -0,0 +1,252 @@
+        return 25 * 1024 * 1024;
+      }
+    };
+
+    ObjectMapper mapper = new DefaultObjectMapper();
+    final IndexIO indexIO = new IndexIO(mapper, config);
+    _index = indexIO.loadIndex(indexDir);
+    QueryableIndexStorageAdapter adapter = new QueryableIndexStorageAdapter(_index);
+
+    // A Sequence "represents an iterable sequence of elements. Unlike normal Iterators however, it doesn't expose
+    // a way for you to extract values from it, instead you provide it with a worker (an Accumulator) and that defines
+    // what happens with the data."
+    final Sequence<Cursor> cursors = adapter.makeCursors(
+        Filters.toFilter(null),
+        _index.getDataInterval().withChronology(ISOChronology.getInstanceUTC()),
+        VirtualColumns.EMPTY,
+        Granularities.ALL,
+        false,
+        null
+    );
+
+    List<Cursor> cursorList = new ArrayList<>();
+    final Sequence<Object> sequence = Sequences.map(cursors, new Function<Cursor, Object>() {
+      @Override
+      public Object apply(@Nullable Cursor cursor) {
+        cursorList.add(cursor);
+        return null;
+      }
+    });
+    sequence.accumulate(null, (accumulated, in) -> null);
+
+    // There should only be one single Cursor for every segment, so there should only be one Cursor in the cursorList
+    Preconditions.checkArgument(cursorList.size() == 1, "There should only be one Cursor in the Sequence.");
+    _cursor = cursorList.get(0);
+
+    _columnNames = new ArrayList<>();
+    _columnNames.addAll(_pinotSchema.getColumnNames());
+    validateColumns(adapter);
+
+    ColumnSelectorFactory columnSelectorFactory = _cursor.getColumnSelectorFactory();
+    _selectors = _columnNames
+        .stream()
+        .map(columnSelectorFactory::makeColumnValueSelector)
+        .collect(Collectors.toList());
+  }
+
+  //@Override
+  public void init(SegmentGeneratorConfig segmentGeneratorConfig) {
+
+  }
+
+  @Override
+  public boolean hasNext() {
+    return !_cursor.isDone();
+  }
+
+  @Override
+  public GenericRow next() throws IOException {
+    return next(new GenericRow());
+  }
+
+  @Override
+  public GenericRow next(GenericRow reuse) {
+    // Only the columns whose names are in the Pinot schema will get processed
+    for (int i = 0; i < _columnNames.size(); i++) {
+      final String columnName = _columnNames.get(i);
+      BaseObjectColumnValueSelector selector = _selectors.get(i);
+      // If the column does not exist in the segment file, skip it.
+      if (selector != null) {
+        FieldSpec fieldSpec = _pinotSchema.getFieldSpecFor(_columnNames.get(i));
+        Object value = selector.getObject();
+        if (value != null && !fieldSpec.isSingleValueField()) {
+          // Multi-valued dimensions in Druid are stored as Arrays.ArrayList (this has been checked)
+          Preconditions.checkArgument(value.getClass() == java.util.Arrays.asList().getClass(), "The multi-valued dimension " + columnName +
```

Comment:
For java, when you do the equality check, you should always use `.equals()` if you are comparing java objects. You can only use `==` for primitive types.